ghost/email-analytics-service/lib/email-analytics-service.js

```diff
@@ -25,72 +56,270 @@ module.exports = class EmailAnalyticsService {
+                this.#fetchScheduledData.canceled = true;
+            } else {
+                this.#fetchScheduledData = null;
             }
         }
-        debug(`fetchLatest: finished in ${Date.now() - startFetch}ms. Fetched ${result.totalEvents} events`);
+    }
+
+    /**
+     * Continues fetching the scheduled events (does not start one). Resets the scheduled event when received 0 events.
+     */
+    async fetchScheduled({maxEvents = Infinity} = {}) {
+        if (!this.#fetchScheduledData || !this.#fetchScheduledData.schedule) {
+            // Nothing scheduled
+            return 0;
+        }
 
-        return result;
+        if (this.#fetchScheduledData.canceled) {
+            // Skip for now
+            this.#fetchScheduledData = null;
+            return 0;
+        }
+
+        let begin = this.#fetchScheduledData.schedule.begin;
+        const end = this.#fetchScheduledData.schedule.end;
+
+        if (this.#fetchScheduledData.lastEventTimestamp && this.#fetchScheduledData.lastEventTimestamp > begin) {
+            // Continue where we left of
+            begin = this.#fetchScheduledData.lastEventTimestamp;
+        }
+
+        if (end < begin) {
+            // Skip for now
+            logging.info('[EmailAnalytics] Skipping fetchScheduled because end is before begin');
+            this.#fetchScheduledData = null;
+            return 0;
+        }
+
+        const count = await this.#fetchEvents(this.#fetchScheduledData, {begin, end, maxEvents});
+        if (count === 0 || this.#fetchScheduledData.canceled) {
+            // Reset the scheduled fetch
+            this.#fetchScheduledData = null;
+        }
+        return count;
     }
 
-    async processEventBatch(events) {
-        const result = new EventProcessingResult();
+    /**
+     * Start fetching analytics and store the data of the progress inside fetchData
+     * @param {FetchData} fetchData
+     * @param {object} options
+     * @param {Date} options.begin
+     * @param {Date} options.end
+     * @param {number} [options.maxEvents] Not a strict maximum. We stop fetching after we reached the maximum AND received at least one event after begin (not equal) to prevent deadlocks.
+     */
+    async #fetchEvents(fetchData, {begin, end, maxEvents = Infinity}) {
+        // Start where we left of, or the last stored event in the database, or start 30 minutes ago if we have nothing available
+        logging.info('[EmailAnalytics] Fetching from ' + begin.toISOString() + ' until ' + end.toISOString() + ' (maxEvents: ' + maxEvents + ')');
+
+        // Store that we started fetching
+        fetchData.running = true;
+        fetchData.lastStarted = new Date();
+        fetchData.lastBegin = begin;
+
+        let lastAggregation = Date.now();
+        let eventCount = 0;
+
+        // We keep the processing result here, so we also have a result in case of failures
+        let processingResult = new EventProcessingResult();
```

Comment:
The way this works has changed. In the past, we kept on creating new `EventProcessingResult`s and passing and merging them around. IMO with the number of events that caused a lot of CPU overhead (array merges + unique filters on those arrays). Instead, all events are appended to one EventProcessingResult. Apart from that `EventProcessingResult` is reset every 5 minutes after aggregating the related emails and members stats. That keeps it from becoming too big + that we have to wait a long time before we start aggregating the data (= stats become visible for the publishers).

Reply:
This is also the reason the processingResult is not returned any longer and why the aggregation moved to the analytics service (instead of the wrapper). Also in case of fatal errors during processing, we can still aggregate the events that we received before that error (otherwise we could have out of date aggregated stats in emails and members).